[docs] Add a tutorial for the pass manager (#3515)
dmlc/tvm: docs/dev/relay_pass_infra.rst

Function-level Passes
^^^^^^^^^^^^^^^^^^^^^

Function-level passes are used to implement various intra function level
optimizations for a given Relay module. It fetches one function at a time from
the function list of a module for optimization and yields a rewritten Relay
function. Most of Relay passes can be classified into this category, such as
common subexpression elimination and inference simplification, etc.

Note that the scope of passes at this level is a Relay function. Therefore, we
cannot add or delete a function through these passes as they are not aware of
the global information.

.. code:: c++
   
    class FunctionPassNode : PassNode {
      PassInfo pass_info;
      runtime::TypedPackedFunc<Function(Function, Module, PassContext)> pass_func;
      Module operator()(const Module& mod, const PassContext& pass_ctx) const final;
      bool SkipFunction(const Function& func) const;
      // Other members/methods are omitted...
    };

``pass_info`` is identical to what we just described in the module pass.
``pass_func`` takes a function for optimization, it also needs a module as we
may use it for reporting errors. A function could be annotated with

> "SkipOptimization" so that it will be ignored during optimization.

Author: Yeah, I think `SkipOptimization` may be a better name. We should probably make the function name `SkipOptimization` as well.

Sequential Passes
^^^^^^^^^^^^^^^^^

SequentialPass is similar to Pytorch nn.Sequential that contains a host of
passes for execution. It also has a disabled list that contains the list of
disabled passes. For example, we may have a configuration which performs all
passes at ``opt_level=3`` with some disabled passes using ``disabled_pass=xx``
provided by ``PassContext``. Now using ``SequentialPass``, we could glob all
passes at ``opt_level=3`` and exclude those in disabled.

.. code:: c++

    class SequentialPassNode : PassNode {
      PassInfo pass_info;
      // Passes need to be executed.
      Array<Pass> passes;
      // Passes are disabled during optimization

>       std::vector<std::string> disabled;

Reviewer: Why do not you use Array<StringImm> ?

Author: Thanks for pointing out. `disabled` shouldn't belong here. We only have it in `PassContext` which is a `Array<tvm::Expr>`

..  Licensed to the Apache Software Foundation (ASF) under one
    or more contributor license agreements.  See the NOTICE file
    distributed with this work for additional information
    regarding copyright ownership.  The ASF licenses this file
    to you under the Apache License, Version 2.0 (the
    "License"); you may not use this file except in compliance
    with the License.  You may obtain a copy of the License at

..    http://www.apache.org/licenses/LICENSE-2.0

..  Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.

Relay Pass Infra
==================================

Relay features a series of optimization passes which improve performance metrics
of models such as mean inference, memory footprint, or power consumption for
specific devices. There are a suite of standard optimizations as well as machine

> learning specific optimizations including constant folding, dead code

Reviewer 2: ```suggestion
learning-specific optimizations including constant folding, dead code
```

elimination, operator layout alteration, and operator fusion, etc. Each of these

> passes are structured as Relay-to-Relay transformations on the abstract syntax

Reviewer 2: ```suggestion
passes is structured as a Relay-to-Relay transformation on the abstract syntax
```

tree (AST) using the analysis result collected during and/or before traversal.

> On one hand, many existing production compilers, such as GCC and LLVM, employ

Reviewer 2: It is confusing to have two paragraphs start with "On one hand" and "on the other" without explicitly setting up the contrast/tension beforehand. That is, there should be a sentence or paragraph first that sets up the tension between these two approaches.

pass managers to effectively manage the execution of passes. Initially managing
passes is straightforward as the number of passes is small, but mature compilers
will contain hundreds of individual passes. Often external users will want to
have custom passes correctly scheduled without having to modify a single
handcrafted pass order.

On the other hand, modern deep learning frameworks, such as Pytorch and MXNet

> Gluon, also have the tendency to enable pass style layer construction

Reviewer 2: ```suggestion
Gluon, also have the tendency to enable pass-style layer construction
```

scheme through `Sequential`_ and `Block`_, respectively. With such constructs,
these modern frameworks are able to conveniently add modules/layers to their
containers and build up neural network easily.

The design of the Relay pass infra is largely inspired by the the hierarchical
pass manager used in LLVM and the block style containers used in the popular

> deep learning frameworks. The major goals of the pass infra include:

Reviewer 2: Are the two paragraphs before this one meant to set up a contrast between the two approaches? If so, this should explain how Relay's pass manager borrows from both.

Author: Thank you for pointing out. I am actually not trying to setup a contrast. Instead, I want to say that we considered the advantages of both other approaches. I will rephrase.

#) enabling better programmatic orchestration of optimizations. This allows
   users to flexibly customize and build their own optimization pipelines.

> #) providing a user friendly way to debug optimization passes.

Reviewer 2: ```suggestion
#) providing a user-friendly way to debug optimization passes.
```

#) alleviating developers from manually and respectively resolving the
   dependencies between passes.

#) simplifying the implementation of new passes for developers. For example, we
   allow users to implement a pass in Python and let the pass infra manipulate
   its execution.

The Design
----------

We focus on ease of extension for users, making it possible for users to quickly
add new passes without loss of backward compatibility. The design contains both
the backend and the frontend. The former implements the main logic of the pass
infra. The latter provides simple APIs for users to interact with, i.e.
allowing users to quickly create their own optimization pipelines.

C++ Backend
~~~~~~~~~~~

We provide a ``PassInfo`` object to contain the basic information needed by

> a pass. ``name`` is the pass name, ``opt_level`` indicates at which optimization

Reviewer 2: Could there be a brief explanation of the opt_level mechanism? Or is it something entirely handled by the pass manager?

level the pass will be enabled, and ``required`` represents the passes that are
required to execute a certain pass (see `include/tvm/relay/transform.h`_ for
more details).

.. code:: c++

    class PassInfoNode : public RelayNode {
      std::string name;
      int opt_level;
      std::vector<std::string> required;
    };

PassContext
^^^^^^^^^^^

``PassContext`` carries useful information for an optimization pass. For
example, it contains the error reporting system so optimization authors can
provide diagnostics about why an optimization fails. ``PassContext`` is also
designed to replace the old ``BuildConfig`` which was used to help users
configure the compilation options, including optimization level and
required/disabled passes, etc. For instance, we may have a configuration which
performs all passes at ``opt_level=3`` with some disabled passes using

> ``disabled_pass=xx`` provided by ``PassContext``. Now  we could glob all passes

Reviewer 2: Is "glob" a technical term used in this context? It seems a bit informal for this context otherwise.

Reviewer 3: I think it stems from the term "globbing" in Bash, but it probably _is_ too informal for the docs.  Maybe "collect" instead?

> at ``opt_level=3`` and exclude those in disabled pass list.

Reviewer 2: ```suggestion
at ``opt_level=3`` and exclude those in the disabled pass list.
```

> This class is designed as a thread local so that users can conveniently use the

Reviewer 2: What does "thread-local" mean in this context?

Author: I will rephrase this. I wanted to say that we used a thread-local store to keep the pass context objects. Users can get the current context in a thread-safe way.

Python ``with`` syntax to perform optimizations under a certain configuration.

> Examples will be provided later to show how we can use both C++ and Python APIs

Reviewer 2: ```suggestion
Examples will be provided later to show how we can use both the C++ and Python APIs
```

to create a compilation pipeline using pass context.

.. code:: c++

    class PassContextNode : public RelayNode {
     public:
      ErrorReporter err_reporter;
      int opt_level{2};
      int fallback_device{static_cast<int>(kDLCPU)};
      tvm::Array<tvm::Expr> required_pass;
      tvm::Array<tvm::Expr> disabled_pass;
    };

.. code:: c++

    class PassContext : public NodeRef {
     public:
      TVM_DLL static PassContext Create();
      TVM_DLL static PassContext Current();
      /* Other fields are omitted. */
    
     private:
      // The entry of a pass context scope.
      TVM_DLL void EnterWithScope();
      // The exit of a pass context scope.
      TVM_DLL void ExitWithScope();
    
      // Classes to get the Python `with` like syntax.
      friend class tvm::With<PassContext>;
    };
   

Pass Constructs
^^^^^^^^^^^^^^^

The pass infra is designed in a hierarchical manner, and it could work at

> various granularity of a Relay program. A pure virtual class ``PassNode`` is

Reviewer 2: ```suggestion
different granularities of Relay programs. A pure virtual class ``PassNode`` is
```

introduced to serve as the base of the different optimization passes. This class

> contains several virtual methods that will be implemented by the

Reviewer 2: ```suggestion
contains several virtual methods that must be implemented by the
```

> subclasses at the module level, function level, or a sequence of passes..

Reviewer 2: ```suggestion
subclasses at the level of modules, functions, or sequences of passes.
```
(grammatical parallelism)

.. code:: c++

    class PassNode : RelayNode {
      virtual PassInfo Info() const = 0;
      virtual Module operator()(const Module& mod
                                const PassContext& pass_ctx) const = 0;
    };

> The functor shows how a pass will be realized, i.e. it always works on a `Relay

Reviewer 2: ```suggestion
The functor shows how a pass must be realized, i.e., it always works on a `Relay
```

> module`_ under a certain context. All passes are designed in a ``Module`` to

Reviewer 2: ```suggestion
module`_ under a certain context. All passes are designed in a ``Module``-to-``Module``
```

> ``Module`` manner. Therefore, optimizations governed by the pass infra will

Reviewer 2: ```suggestion
manner. Therefore, optimizations governed by the pass infra will
```
(combined with the previous suggestion)

always update the whole module.

> Several subclasses are created to implement different types of optimization

Reviewer 2: ```suggestion
Several subclasses have been created to implement different types of optimization
```

> passes, e.g. function-level passes, module-level passes, and sequential passes.

Reviewer 2: ```suggestion
passes, e.g., function-level passes, module-level passes, and sequential passes.
```
Also it should be explained what "sequential passes" means somewhere. (I presume it means a sequence of passes; is that right?)

Author: yes, it is right.

Each subclass itself could act as a pass manager. For instance, they could glob

> the require passes and execute them or build a dependency graph based on the

Reviewer 2: ```suggestion
the required passes and execute them or build a dependency graph based on the
```

> given meta data. The full definition of them could be found in

Reviewer 2: ```suggestion
given metadata. The full definition of them could be found in
```

`src/relay/pass/pass_manager.cc`_

Module-Level Passes
^^^^^^^^^^^^^^^^^^^

Module level passes are geared mainly for global and inter-procedural

> optimizations (IPO), which is similar to the module pass used in LLVM. Some

Reviewer 2: ```suggestion
optimizations (IPO), which are similar to the module pass used in LLVM. Some
```

> typical passes in Relay that need the global picture of a module, such as to

Reviewer 2: ```suggestion
typical passes in Relay that need the global picture of a module, such as
```

> a normal form conversion and lambda lifting, etc, fall in this set. At this

Reviewer 2: ```suggestion
A-normal form conversion and lambda lifting, etc., fall into this set. At this
```

optimizations for a given Relay module. It fetches one function at a time from
the function list of a module for optimization and yields a rewritten Relay
function. Most of Relay passes can be classified into this category, such as
common subexpression elimination and inference simplification, etc.

Note that the scope of passes at this level is a Relay function. Therefore, we
cannot add or delete a function through these passes as they are not aware of
the global information.

.. code:: c++
   
    class FunctionPassNode : PassNode {
      PassInfo pass_info;
      runtime::TypedPackedFunc<Function(Function, Module, PassContext)> pass_func;
      Module operator()(const Module& mod, const PassContext& pass_ctx) const final;
      bool SkipFunction(const Function& func) const;
      // Other members/methods are omitted...
    };

``pass_info`` is identical to what we just described in the module pass.
``pass_func`` takes a function for optimization, it also needs a module as we
may use it for reporting errors. A function could be annotated with
"SkipOptimization" so that it will be ignored during optimization.

Sequential Passes
^^^^^^^^^^^^^^^^^

> SequentialPass is similar to Pytorch nn.Sequential that contains a host of

Reviewer 2: We probably want `code::` annotations for terms like `SequentialPass` or `nn.Sequential` (just to be consistent)

indicates that no prerequisite is required for this pass. Otherwise, the pass
developer has to identify and list them.

Meanwhile, a pass API endpoint is registered with the name
``relay._transform.FoldConstant``. This pass, therefore, becomes an entry in the
registry that can be accessed by both C++ (e.g. the ``GetPass`` above) and
Python when needed.

.. code:: c++

    namespace transform {

    Pass FoldConstant() {
      runtime::TypedPackedFunc<Function(Function, Module, PassContext)> pass_func =
        [=](Function f, Module m, PassContext pc) {
          return Downcast<Function>(FoldConstant(f));
      };
      return CreateFunctionPass(pass_func, 2, "FoldConstant", {});
    }

    TVM_REGISTER_API("relay._transform.FoldConstant")
    .set_body_typed(FoldConstant);

    }  // namespace transform

To allow other C++ modules to apply this pass, we declare a free function in
`include/tvm/relay/transform.h`_ as the following:

.. code:: c++

    TVM_DLL Pass FoldConstant();

Python Frontend
~~~~~~~~~~~~~~~

Only some simple APIs are needed for the frontend side. For example, we can
provide users the following APIs to create and execute a pass (full

> implementation is provided in `python/tvm/relay/transform.py`_). The backend

Reviewer 2: I personally think "back-end" and "front-end" should be hyphenated but I don't know if everyone shares this opinion

Reviewer 3: I think they're common enough words now that we can drop the hyphen.

Author: I will probably just keep them because it looks that they are also commonly used, including academic papers.

function.

.. code:: python

    mod = relay.Module()
    mod = module_pass(mod)

Correspondingly, we also offer such functionality for ``function_pass``. For
instance, an example function-level pass could be written as the following:

.. code:: python

    @relay.transform.function_pass(opt_level=1)
    class TestReplaceFunc:
       def __init__(self, new_func):
          self.new_func = new_func
          def transform_function(self, func, mod, ctx):
             # Just for demo purposes
             # Transform func to new_func
             return self.new_func

    x = relay.var("x", shape=(10, 20))
    f1 = relay.Function([x], x)
    f2 = relay.Function([x], relay.log(x))
    # fpass is now a special pass that replaces every
    # function to f1
    fpass = TestReplaceFunc(f1)
    # Now every function in input_mod is replaced by f1
    res_mod = fpass(input_mod)

Alternatively, users can also directly register a pass without using the
decorators and then invoke it. Let's use ``Sequential`` to demo this scenario.

Python Sequential Example
^^^^^^^^^^^^^^^^^^^^^^^^^

This example not only illustrates how users can directly create a sequential

> pass using Python APIs (this could be applied to module- and function- level

Reviewer 2: ```suggestion
pass using Python APIs (this could be applied to module- and function-level
```

^^^^^^^^^^^

Python frontend provides a wrapper for the ``PassContext`` to enable the
``with`` syntax by overriding ``__enter__`` and ``__exit__``. A ``current``
static method is offered for users to get the context that is in use under
a certain scope.

.. code:: python

    @register_relay_node
    class PassContext(RelayNode):
        def __enter__(self):
            _transform.EnterPassContext(self)
            return self
    
        def __exit__(self, ptype, value, trace):
            _transform.ExitPassContext(self)
    
        @staticmethod
        def current():
            """Return the current pass context."""
            return _transform.GetCurrentPassContext()

A ``PassContext`` object can be instantiated through the ``build_config`` API
which was used by Relay to configure the compilation options, including the
optimization level, fallback device for heterogeneous execution, and
required/disabled passes.

Pass Objects
^^^^^^^^^^^^

``Pass`` is the base class of all pass objects. All methods here are just simple
wrappers that were implemented in the backend. They are defined for users to
conveniently interact with the base class in Python. Only a functor is defined

> to invoke the execution of a pass.

Reviewer 3: This sentence is a little vague.  What functor?

Author: Rephrase to ``__call__``

..  Licensed to the Apache Software Foundation (ASF) under one
    or more contributor license agreements.  See the NOTICE file
    distributed with this work for additional information
    regarding copyright ownership.  The ASF licenses this file
    to you under the Apache License, Version 2.0 (the
    "License"); you may not use this file except in compliance
    with the License.  You may obtain a copy of the License at

..    http://www.apache.org/licenses/LICENSE-2.0

..  Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.

Relay Pass Infra
==================================

Relay features a series of optimization passes which improve performance metrics
of models such as mean inference, memory footprint, or power consumption for

> specific devices. There are a suite of standard optimizations as well as machine

Reviewer 2: ```suggestion
specific devices. There is a suite of standard optimizations as well as machine
```

learning-specific optimizations including constant folding, dead code
elimination, operator layout alteration, and operator fusion, etc. Each of these

> passes are structured as a Relay-to-Relay transformation on the abstract syntax

Reviewer 2: ```suggestion
passes is structured as a Relay-to-Relay transformation on the abstract syntax
```

tree (AST) using the analysis result collected during and/or before traversal.

> However, as Relay evolves quickly, the need of a more systematic and efficient

Reviewer 2: ```suggestion
However, as Relay evolves quickly, the need for a more systematic and efficient
```

> way to manage these passes becomes imperative. This doc describes the design of

Reviewer 2: ```suggestion
way to manage these passes is becoming apparent. This doc describes the design of
```

Reviewer 2: I changed "imperative" to "apparent" because needs are inherently imperative

scheme through `Sequential`_ and `Block`_, respectively. With such constructs,
these modern frameworks are able to conveniently add modules/layers to their
containers and build up neural networks easily.

The design of the Relay pass infra is largely inspired by the the hierarchical
pass manager used in LLVM and the block-style containers used in the popular
deep learning frameworks. The major goals of the pass infra include:

#) enabling better programmatic orchestration of optimizations. This allows
   users to flexibly customize and build their own optimization pipelines.

#) providing a user-friendly way to debug optimization passes.

#) alleviating developers from manually and respectively resolving the
   dependencies between passes.

#) simplifying the implementation of new passes for developers. For example, we
   allow users to implement a pass in Python and let the pass infra manipulate
   its execution.

The Design
----------

We focus on ease of extension for users, making it possible for users to quickly
add new passes without loss of backward compatibility. The design contains both
the backend and the frontend. The former implements the main logic of the pass

> infra. The latter provides simple APIs for users to interact with, i.e.

Reviewer 2: ```suggestion
infra. The latter provides simple APIs for users to interact with, i.e.,
```

allowing users to quickly create their own optimization pipelines.

C++ Backend
~~~~~~~~~~~

We provide a ``PassInfo`` object to contain the basic information needed by
a pass. ``name`` is the pass name, ``opt_level`` indicates at which optimization
level the pass will be enabled, and ``required`` represents the passes that are
required to execute a certain pass (see `include/tvm/relay/transform.h`_ for
more details). For example, during registration of a pass (will be covered in

> later), the pass developers can specify the name of the pass, the optimiztion

Reviewer 3: ```suggestion
later), the pass developers can specify the name of the pass, the optimization
```

Essentially, we sequentially execute each pass in a sequential pass using the
order that they were appended to the pass list.

.. code:: c++

    Module SequentialNode::operator()(const Module& module,
                                      const PassContext& pass_ctx) const {
      Module mod = module;
      for (const Pass& pass : passes) {
        CHECK(pass.defined()) << "Found undefined pass for optimization.";
        const PassInfo& pass_info = pass->Info();
        if (!PassEnabled(pass_info))  continue;
        for (const auto& it : pass_info->required) {
          const auto* name = it.as<tvm::ir::StringImm>();
          CHECK(name);
          mod = GetPass(name->value)(mod, pass_ctx);
        }
        mod = pass(mod, pass_ctx);
      }
      return mod;
    }

Upon the invocation of a pass, we first check if this pass is enabled. This is
done by first checking if the pass is explicitly disabled by a user, followed by
inspecting if it is specified as a required pass by the user. If it is still
undetermined whether this pass is enabled, its ``opt_level`` will be checked.
This pass will be enabled and therefore executed only when its optimization

> level not is less than the configured optimization level in the pass context.

Reviewer 3: ```suggestion
level is not less than the configured optimization level in the pass context.
```
